README.md
# Ruby NYPL Platform API Client

Simple client for querying the NYPL Platform API.

## Version

> 1.0.0

## Using

1. Configure the client with Platform API credentials
2. Make requests

### Configuration

Example configuration:

```ruby
require 'nypl_platform_api_client'

client = NyplPlatformApiClient.new({
  base_url: "https://platform.nypl.org/api/v0.1/", # Defaults to ENV['PLATFORM_API_BASE_URL']
  client_id: "client-id", # Defaults to ENV['NYPL_OAUTH_ID']
  client_secret: "client-secret", # Defaults to ENV['NYPL_OAUTH_SECRET']
  oauth_url: "https://isso.nypl.org" # Defaults to ENV['NYPL_OAUTH_URL'],
  log_level: "debug" # Defaults to 'info'
})
```

### Requests

Example GET:

```ruby
bib = client.get 'bibs/sierra-nypl/12082323'
```

Example POST:

```ruby
new_job = client.post 'jobs', '', { 'Content-Type' => 'text/plain' }
```

## Contributing

This repo uses a single, versioned `master` branch.

 * Create feature branch off `master`
 * Compute next logical version and update `README.md`, `CHANGELOG.md`, & `nypl_platform_api_client.gemspec`
 * Create PR against `master`
 * After merging the PR, git tag `master` with new version number.

### Updating gem

After merging to `master`, push the updated gem to rubygems.org:

```
gem build nypl_platform_api_client.gemspec
gem push nypl_platform_api_client-[version].gem
```

See [this guide](https://guides.rubygems.org/make-your-own-gem/) for additional help.

## Testing

```
bundle exec rspec
```


CHANGELOG.md
# Changelog

## v1.0.0

Initial push with support for:
 - Arbitrary GET and POST requests
 - Optional authentication


spec/spec_helper.rb
require 'json'
require 'nypl_log_formatter'

require_relative '../lib/nypl_platform_api_client'

ENV['LOG_LEVEL'] ||= 'error'
ENV['APP_ENV'] = 'test'

def load_fixture (file)
  JSON.parse File.read("./spec/fixtures/#{file}")
end


spec/nypl_platform_api_client_spec.rb
require 'spec_helper'
require 'webmock/rspec'

describe NyplPlatformApiClient do
  before(:each) do
    ENV['PLATFORM_API_BASE_URL'] = 'https://example.com/api/v0.1/'
    ENV['NYPL_OAUTH_ID'] = Base64.strict_encode64 'fake-client'
    ENV['NYPL_OAUTH_SECRET'] = Base64.strict_encode64 'fake-secret'
    ENV['NYPL_OAUTH_URL'] = 'https://isso.example.com/'

    stub_request(:post, "#{ENV['NYPL_OAUTH_URL']}oauth/token").to_return(status: 200, body: '{ "access_token": "fake-access-token" }')
    stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}bibs/sierra-nypl/12082323").to_return(status: 200, body: File.read('./spec/fixtures/bib.json'))
  end

  describe :config do
    it "should throw error if api base url unset" do
      ENV['PLATFORM_API_BASE_URL'] = nil
      expect { NyplPlatformApiClient.new }.to raise_error(NyplPlatformApiClientError)
    end

    it "should throw error if client id unset" do
      ENV['NYPL_OAUTH_ID'] = nil
      expect { NyplPlatformApiClient.new }.to raise_error(NyplPlatformApiClientError)
    end

    it "should throw error if client secret unset" do
      ENV['NYPL_OAUTH_SECRET'] = nil
      expect { NyplPlatformApiClient.new }.to raise_error(NyplPlatformApiClientError)
    end

    it "should throw error if oauth url unset" do
      ENV['NYPL_OAUTH_URL'] = nil
      expect { NyplPlatformApiClient.new }.to raise_error(NyplPlatformApiClientError)
    end

    it "should allow configuration via constructor" do
      ENV['PLATFORM_API_BASE_URL'] = nil
      ENV['NYPL_OAUTH_ID'] = nil
      ENV['NYPL_OAUTH_SECRET'] = nil
      ENV['NYPL_OAUTH_URL'] = nil

      config = NyplPlatformApiClient.new({
        base_url: 'https://example.com/api/v0.1/',
        client_id: 'client-id',
        client_secret: 'client-secret',
        oauth_url: 'https://isso.example.com'
      }).instance_variable_get(:@config)

      expect(config).to be_a(Hash)
      expect(config[:base_url]).to eq('https://example.com/api/v0.1/')
      expect(config[:client_id]).to eq('client-id')
    end

    it "should prefer constructor config over ENV variables" do
      client = NyplPlatformApiClient.new({
        client_id: 'client-id-via-constructor-config'
      })
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:client_id]).to eq('client-id-via-constructor-config')
    end

    it "should set default log_level to 'info'" do
      client = NyplPlatformApiClient.new
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('info')
    end

    it "should allow log_level override via constructor (only)" do
      ENV['LOG_LEVEL'] = 'debug'
      client = NyplPlatformApiClient.new
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('info')

      client = NyplPlatformApiClient.new(log_level: 'debug')
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('debug')
    end
  end

  describe :parse_http_options do
    it "should assume common defaults" do
      options = NyplPlatformApiClient.new.send :parse_http_options, {}

      expect(options[:authenticated]).to eq(true)
      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers].keys.size).to eq(0)
    end

    it "should allow authentication override" do
      options = NyplPlatformApiClient.new.send :parse_http_options, { authenticated: false }

      expect(options[:authenticated]).to eq(false)
    end

    it "should allow custom Content-Type" do
      options = NyplPlatformApiClient.new.send :parse_http_options, { headers: { 'Content-Type' => 'text/plain' } }

      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers]['Content-Type']).to be_a(String)
      expect(options[:headers]['Content-Type']).to eq('text/plain')
    end

    it "should allow extra header" do
      options = NyplPlatformApiClient.new.send :parse_http_options, { headers: { 'X-My-Header': 'header value' } }

      expect(options[:authenticated]).to eq(true)
      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers]['X-My-Header']).to eq('header value')
    end
  end

  describe :authentication do

    it "should authenticate by default" do
      client = NyplPlatformApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint with authentication:
      expect(client.get('bibs/sierra-nypl/12082323')).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_a(String)
      expect(client.instance_variable_get(:@access_token)).to eq('fake-access-token')
    end

    it "should authenticate when calling with :authenticated => true" do
      client = NyplPlatformApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint with authentication:
      expect(client.get('bibs/sierra-nypl/12082323', authenticated: true)).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_a(String)
      expect(client.instance_variable_get(:@access_token)).to eq('fake-access-token')
    end

    it "should NOT authenticate when calling with :authenticated => false" do
      client = NyplPlatformApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint without authentication:
      expect(client.get('bibs/sierra-nypl/12082323', authenticated: false)).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_nil
    end
  end

  describe :responses do
    it "should auto parse JSON if 200 response" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 200, body: '{ "foo": "bar" }' )

      resp = NyplPlatformApiClient.new.get('some-path')
      expect(resp).to be_a(Hash)
      expect(resp['foo']).to eq('bar')
    end

    it "should auto parse JSON if 404 response" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 404, body: '{ "foo": "bar" }' )

      resp = NyplPlatformApiClient.new.get('some-path')
      expect(resp).to be_a(Hash)
      expect(resp['foo']).to eq('bar')
    end

    it "should consider non-200, non-404 as errors" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 300, body: '{ "foo": "bar" }' )
      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientError)

      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 500, body: '{ "foo": "bar" }' )
      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientError)
    end

    it "should throw NyplPlatformApiClientError if response is not valid json" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 200, body: '' )
      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientError)

      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 200, body: '{ "foo": "bar' )
      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientError)
    end

    it "should throw NyplPlatformApiClientError for 401" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 401, body: '' )

      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientTokenError)
    end

    it "should throw NyplPlatformApiClientError for 500" do
      stub_request(:get, "#{ENV['PLATFORM_API_BASE_URL']}some-path").to_return(status: 500, body: '' )

      expect { NyplPlatformApiClient.new.get('some-path') }.to raise_error(NyplPlatformApiClientError)
    end
  end
end


lib/nypl_platform_api_client.rb
require 'net/http'
require 'net/https'
require 'uri'

require_relative 'errors'

class NyplPlatformApiClient
  def initialize(config = {})
    @config = {
      base_url: ENV['PLATFORM_API_BASE_URL'],
      client_id: ENV['NYPL_OAUTH_ID'],
      client_secret: ENV['NYPL_OAUTH_SECRET'],
      oauth_url: ENV['NYPL_OAUTH_URL'],
      log_level: 'info'
    }.merge config

    raise NyplPlatformApiClientError.new 'Missing config: neither config.base_url nor ENV.PLATFORM_API_BASE_URL are set' unless @config[:base_url]
    raise NyplPlatformApiClientError.new 'Missing config: neither config.client_id nor ENV.NYPL_OAUTH_ID are set' unless @config[:client_id]
    raise NyplPlatformApiClientError.new 'Missing config: neither config.client_secret nor ENV.NYPL_OAUTH_SECRET are set ' unless @config[:client_secret]
    raise NyplPlatformApiClientError.new 'Missing config: neither config.oauth_url nor ENV.NYPL_OAUTH_URL are set ' unless @config[:oauth_url]
  end

  def get (path, options = {})
    options = parse_http_options options

    authenticate! if options[:authenticated]

    uri = URI.parse("#{@config[:base_url]}#{path}")

    logger.debug "NyplPlatformApiClient: Getting from platform api", { uri: uri }

    begin
      request = Net::HTTP::Get.new(uri)

      # Add bearer token header
      request["Authorization"] = "Bearer #{@access_token}" if options[:authenticated]

      # Execute request:
      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme === 'https') do |http|
        http.request(request)
      end
    rescue Exception => e
      raise NyplPlatformApiClientError.new(e), "Failed to GET #{path}: #{e.message}"
    end

    logger.debug "NyplPlatformApiClient: Got platform api response", { code: response.code, body: response.body }

    parse_json_response response
  end

  def post (path, body, options = {})
    options = parse_http_options options

    # Default to POSTing JSON unless explicitly stated otherwise
    options[:headers]['Content-Type'] = 'application/json' unless options[:headers]['Content-Type']

    authenticate! if options[:authenticated]

    uri = URI.parse("#{@config[:base_url]}#{path}")

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme === 'https'

    begin
      request = Net::HTTP::Post.new(uri.path, 'Content-Type' => options[:headers]['Content-Type'])
      request.body = body.to_json

      logger.debug "NyplPlatformApiClient: Posting to platform api", { uri: uri, body: body }

      # Add bearer token header
      request['Authorization'] = "Bearer #{@access_token}" if options[:authenticated]

      # Execute request:
      response = http.request(request)
    rescue Exception => e
      raise NyplPlatformApiClientError.new(e), "Failed to POST to #{path}: #{e.message}"
    end

    logger.debug "NyplPlatformApiClient: Got platform api response", { code: response.code, body: response.body }

    parse_json_response response
  end

  private

  def parse_http_options (_options)
    options = {
      authenticated: true
    }.merge _options

    options[:headers] = {
    }.merge(_options[:headers] || {})
      .transform_keys(&:to_s)

    options
  end

  def parse_json_response (response)
    # Among NYPL services, these are the only non-error statuses with useful JSON payloads:
    if ['200', '404'].include? response.code
      begin
        JSON.parse(response.body)
      rescue => e
        raise NyplPlatformApiClientError, "Error parsing response (#{response.code}): #{response.body}"
      end
    elsif response.code == "401"
      # Likely an expired access-token; Wipe it for next run
      # TODO: Implement token refresh
      @access_token = nil
      raise NyplPlatformApiClientTokenError.new("Got a 401: #{response.body}")
    else
      raise NyplPlatformApiClientError, "Error interpretting response (#{response.code}): #{response.body}"
    end
  end

  # Authorizes the request.
  def authenticate!
    # NOOP if we've already authenticated
    return nil if ! @access_token.nil?

    logger.debug "NyplPlatformApiClient: Authenticating with client_id #{@config[:client_id]}"

    uri = URI.parse("#{@config[:oauth_url]}oauth/token")
    request = Net::HTTP::Post.new(uri)
    request.basic_auth(@config[:client_id], @config[:client_secret])
    request.set_form_data(
      "grant_type" => "client_credentials"
    )

    req_options = {
      use_ssl: uri.scheme == "https",
      request_timeout: 500
    }

    response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
      http.request(request)
    end

    if response.code == '200'
      @access_token = JSON.parse(response.body)["access_token"]
    else
      nil
    end
  end

  def logger
    @logger ||= NyplLogFormatter.new(STDOUT, level: @config[:log_level])
  end
end
